src/photon/photon_algorithm.c

```diff
@@ -168,78 +182,144 @@ void fetch_missing_dependencies(local_scheduler_state *state,
+    DL_DELETE(algorithm_state->dispatch_task_queue, dispatched_task);
 
-void run_task_immediately(local_scheduler_state *state,
-                          scheduling_algorithm_state *algorithm_state,
-                          task_spec *spec,
-                          bool from_global_scheduler) {
-  /* Get the last available worker in the available worker queue. */
-  int *worker_index = (int *) utarray_back(algorithm_state->available_workers);
-  /* Tell the available worker to execute the task. */
-  assign_task_to_worker(state, spec, *worker_index, from_global_scheduler);
-  /* Remove the available worker from the queue and free the struct. */
-  utarray_pop_back(algorithm_state->available_workers);
+    /* Get the last available worker in the available worker queue. */
+    int *worker_index =
+        (int *) utarray_back(algorithm_state->available_workers);
+    /* Tell the available worker to execute the task. */
+    assign_task_to_worker(state, dispatched_task->spec, *worker_index);
+    /* Remove the available worker from the queue and free the struct. */
+    utarray_pop_back(algorithm_state->available_workers);
+    free_task_spec(dispatched_task->spec);
+    free(dispatched_task);
+  }
 }
 
-void queue_task_locally(local_scheduler_state *state,
-                        scheduling_algorithm_state *algorithm_state,
-                        task_spec *spec,
-                        bool from_global_scheduler) {
+/**
+ * A helper function to allocate a queue entry for a task specification and
+ * push it onto a generic queue.
+ *
+ * @param task_queue A pointer to a task queue. NOTE: Because we are using
+ *        utlist.h, we must pass in a pointer to the queue we want to append
+ *        to. If we passed in the queue itself and the queue was empty, this
+ *        would append the task to a queue that we don't have a reference to.
+ * @param spec The task specification to queue.
+ * @return Void.
+ */
+void queue_task(local_scheduler_state *state,
+                task_queue_entry **task_queue,
+                task_spec *spec) {
   /* Copy the spec and add it to the task queue. The allocated spec will be
    * freed when it is assigned to a worker. */
   task_queue_entry *elt = malloc(sizeof(task_queue_entry));
   elt->spec = (task_spec *) malloc(task_spec_size(spec));
   memcpy(elt->spec, spec, task_spec_size(spec));
-  elt->from_global_scheduler = from_global_scheduler;
-  DL_APPEND(algorithm_state->task_queue, elt);
-  if (!from_global_scheduler && state->db != NULL) {
+  DL_APPEND((*task_queue), elt);
+
+  /* The task has been added to a local scheduler queue. Add the task to the
+   * task table to notify others that we have queued it. */
+  if (state->db != NULL) {
     task *task =
-        alloc_task(spec, TASK_STATUS_SCHEDULED, get_db_client_id(state->db));
+        alloc_task(spec, TASK_STATUS_QUEUED, get_db_client_id(state->db));
     task_table_add_task(state->db, task, (retry_info *) &photon_retry, NULL,
```

Comment:
This shouldn't always be a `task_table_add_task`, right? Sometimes the task is already there and it should be a `task_table_update`.